packages/@react-aria/utils/src/useLoadMore.ts

```diff
@@ -0,0 +1,78 @@
+import {useEvent} from './useEvent';
+// eslint-disable-next-line rulesdir/useLayoutEffectRule
+import {useLayoutEffect} from './useLayoutEffect';
+
+export interface LoadMoreProps {
+  /** Whether data is currently being loaded. */
+  isLoading?: boolean,
+  /** Handler that is called when more items should be loaded, e.g. while scrolling near the bottom.  */
+  onLoadMore?: () => void,
+  /**
+   * The amount of offset from the bottom of your scrollable region that should trigger load more.
+   * Uses a percentage value relative to the scroll body's client height. Load more is then triggered
+   * when your current scroll position's distance from the bottom of the currently loaded list of items is less than
+   * or equal to the provided value. (e.g. 1 = 100% of the scroll region's height).
+   * @default 1
+   */
+  scrollOffset?: number
+}
+
+export function useLoadMore(props: LoadMoreProps, ref: RefObject<HTMLElement | null>) {
+  let {isLoading, onLoadMore, scrollOffset = 1} = props;
+
+  // Handle scrolling, and call onLoadMore when nearing the bottom.
+  let isLoadingRef = useRef(isLoading);
+  let prevProps = useRef(props);
+  let onScroll = useCallback(() => {
+    if (ref.current && !isLoadingRef.current && onLoadMore) {
+      let shouldLoadMore = ref.current.scrollHeight - ref.current.scrollTop - ref.current.clientHeight < ref.current.clientHeight * scrollOffset;
+
+      if (shouldLoadMore) {
+        isLoadingRef.current = true;
+        onLoadMore();
+      }
+    }
+  }, [onLoadMore, ref, scrollOffset]);
+
+  useLayoutEffect(() => {
+    // Only update isLoadingRef if props object actually changed,
+    // not if a local state change occurred.
+    if (props !== prevProps.current) {
+      isLoadingRef.current = isLoading;
+      prevProps.current = props;
+    }
+
+    // TODO: this actually calls loadmore twice in succession on intial load because after the first load
+    // the scrollable element hasn't yet recieved its new height with the newly loaded items... Because of RAC collection needing two renders?
+    // Using ref.current.clientElementCount doesn't work either because the scrollable body may be the Table resize container (which contains the table, header and body: 3 children always)
+    // or the Table (which contains the header and body so 2 children) or the virtualizer in RSP (just contains the rows so has a variable count)
+    let shouldLoadMore = ref?.current
+      && !isLoadingRef.current
+      && onLoadMore
+      && ref.current.clientHeight === ref.current.scrollHeight;
+
+    if (shouldLoadMore) {
+      isLoadingRef.current = true;
+      onLoadMore?.();
+    }
+
+  }, [isLoading, onLoadMore, props, ref]);
+
+  // TODO: maybe this should still just return scroll props?
+  // Test against case where the ref isn't defined when this is called
+  // Think this was a problem when trying to attach to the scrollable body of the table in OnLoadMoreTableBodyScroll
+  useEvent(ref, 'scroll', onScroll);
```

Comment:
Bit surprised that this is an issue, double checking if useEvent just doesn't quite work if the ref provided to it isn't immediately attached to an actual node

Reply:
At the point this useEvent is called, Table hasn't rendered anything within it
<img width="280" alt="image" src="https://github.com/adobe/react-spectrum/assets/9661127/9ef70d9f-f8f0-4385-bb8f-77f0c5fcfb57">
so we hit https://github.com/adobe/react-spectrum/blob/2964e9aafc281888cd65da5bd77512e7f999bd84/packages/%40react-aria/utils/src/useEvent.ts#L26-L28. When the Table's internals are rendered after the collection is processed, nothing in the useEffect def array has changed to cause it to trigger again, hence addEventListener is never called on the table body's ref

Reply:
For future reference, we may be moving useLoadMore back into the RAC components in the future so that it can access the component's collection for https://github.com/adobe/react-spectrum/pull/6494#discussion_r1669614731. When this happens, we can always place it in TableBody itself and walk up to find the first scrollable element and set that as the scroll ref, thus allowing for the body/table/ or resizable table container to be the scrollable body